File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/BubbleHandler.java
package closer.vlllage.com.closer.handler.bubble;

import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;

import java.util.List;

import closer.vlllage.com.closer.pool.PoolMember;

public class BubbleHandler extends PoolMember {

    private MapBubbleView.OnMapBubbleClickListener onClickListener;
    private MapBubbleMenuView.OnMapBubbleMenuItemClickListener onMenuItemClickListener;
    private MapBubbleSuggestionView.MapBubbleSuggestionClickListener onMapBubbleSuggestionClickListener;
    private MapBubbleEventView.MapBubbleEventClickListener onMapBubbleEventClickListener;
    private MapBubblePhysicalGroupView.MapBubblePhysicalGroupClickListener onMapBubblePhysicalGroupClickListener;

    private final BubbleMapLayer bubbleMapLayer = new BubbleMapLayer();
    private final BubbleProxyLayer bubbleProxyLayer = new BubbleProxyLayer(bubbleMapLayer);

    public void attach(ViewGroup bubbleMapLayerLayout,
                       MapBubbleView.OnMapBubbleClickListener onClickListener,
                       MapBubbleMenuView.OnMapBubbleMenuItemClickListener onMenuItemClickListener,
                       MapBubbleEventView.MapBubbleEventClickListener onMapBubbleEventClickListener,
                       MapBubbleSuggestionView.MapBubbleSuggestionClickListener onMapBubbleSuggestionClickListener,
                       MapBubblePhysicalGroupView.MapBubblePhysicalGroupClickListener onMapBubblePhysicalGroupClickListener) {
        this.onClickListener = onClickListener;
        this.onMenuItemClickListener = onMenuItemClickListener;
        this.onMapBubbleEventClickListener = onMapBubbleEventClickListener;
        this.onMapBubbleSuggestionClickListener = onMapBubbleSuggestionClickListener;
        this.onMapBubblePhysicalGroupClickListener = onMapBubblePhysicalGroupClickListener;
        bubbleMapLayer.attach(bubbleMapLayerLayout, getBubbleView());
    }

    public void attach(GoogleMap map) {
        bubbleMapLayer.attach(map);
    }

    private BubbleMapLayer.BubbleView getBubbleView() {
        return new BubbleMapLayer.BubbleView() {
            @Override
            public View createView(ViewGroup view, MapBubble mapBubble) {
                switch (mapBubble.getType()) {
                    case PROXY:
                        return $(MapBubbleProxyView.class).from(view, mapBubble, null);
                    case MENU:
                        return $(MapBubbleMenuView.class).from(view, mapBubble, onMenuItemClickListener);
                    case SUGGESTION:
                        return $(MapBubbleSuggestionView.class).from(view, mapBubble, onMapBubbleSuggestionClickListener);
                    case EVENT:
                        return $(MapBubbleEventView.class).from(view, mapBubble, onMapBubbleEventClickListener);
                    case PHYSICAL_GROUP:
                        return $(MapBubblePhysicalGroupView.class).from(view, mapBubble, onMapBubblePhysicalGroupClickListener);
                    default:
                        return $(MapBubbleView.class).from(view, mapBubble, onClickListener);
                }
            }

            @Override
            public void updateView(MapBubble mapBubble) {
                switch (mapBubble.getType()) {
                    case STATUS:
                        $(MapBubbleView.class).update(mapBubble.getView(), mapBubble);
                        break;
                }
            }
        };
    }

    public void add(final MapBubble mapBubble) {
        bubbleProxyLayer.add(mapBubble);
    }

    public void replace(List<MapBubble> mapBubbles) {
        bubbleProxyLayer.replace(mapBubbles);
    }

    public void update() {
        bubbleProxyLayer.update();
    }

    public void update(MapBubble mapBubble) {
        bubbleProxyLayer.update(mapBubble);
    }

    public void move(MapBubble mapBubble, LatLng latLng) {
        bubbleProxyLayer.move(mapBubble, latLng);
    }

    public void updateDetails(MapBubble mapBubble) {
        bubbleProxyLayer.updateDetails(mapBubble);
    }

    public void remove(MapBubble mapBubble) {
        bubbleProxyLayer.remove(mapBubble);
    }

    public boolean remove(BubbleMapLayer.RemoveCallback callback) {
        return bubbleProxyLayer.remove(callback);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/BubbleHandler.kt
package closer.vlllage.com.closer.handler.bubble

import android.view.View
import android.view.ViewGroup
import closer.vlllage.com.closer.handler.map.MapHandler
import at.bluesource.choicesdk.maps.common.LatLng
import at.bluesource.choicesdk.maps.common.Map
import closer.vlllage.com.closer.databinding.MapBubblePhysicalGroupBinding
import com.queatz.on.On

class BubbleHandler constructor(private val on: On) {

    private var onClickListener: ((MapBubble) -> Unit)? = null
    private lateinit var onMenuItemClickListener: OnMapBubbleMenuItemClickListener
    private lateinit var onMapBubbleSuggestionClickListener: MapBubbleSuggestionClickListener
    private lateinit var onMapBubbleEventClickListener: MapBubbleEventClickListener
    private lateinit var onMapBubblePhysicalGroupClickListener: MapBubblePhysicalGroupClickListener

    private val bubbleMapLayer = BubbleMapLayer(on)
    private val bubbleProxyLayer = BubbleProxyLayer(bubbleMapLayer) { on<MapHandler>().visibleRegion }

    private val bubbleView: BubbleMapLayer.BubbleView
        get() = object : BubbleMapLayer.BubbleView {
            override fun createView(view: ViewGroup?, mapBubble: MapBubble): View {
                return when (mapBubble.type) {
                    BubbleType.PROXY -> on<MapBubbleProxyView>().from(view!!, mapBubble) { proxiedMapBubble ->
                        when (proxiedMapBubble.type) {
                            BubbleType.STATUS -> onClickListener!!.invoke(proxiedMapBubble)
                            BubbleType.EVENT -> onMapBubbleEventClickListener.invoke(proxiedMapBubble)
                            BubbleType.PHYSICAL_GROUP -> onMapBubblePhysicalGroupClickListener.invoke(proxiedMapBubble)
                            BubbleType.SUGGESTION -> onMapBubbleSuggestionClickListener.invoke(proxiedMapBubble)
                        }
                    }
                    BubbleType.MENU -> on<MapBubbleMenuView>().from(view!!, mapBubble, onMenuItemClickListener)
                    BubbleType.SUGGESTION -> on<MapBubbleSuggestionView>().from(view!!, mapBubble, onMapBubbleSuggestionClickListener)
                    BubbleType.EVENT -> on<MapBubbleEventView>().from(view!!, mapBubble, onMapBubbleEventClickListener)
                    BubbleType.PHYSICAL_GROUP -> on<MapBubblePhysicalGroupView>().from(view!!, mapBubble, onMapBubblePhysicalGroupClickListener).root
                    else -> on<MapBubblePhysicalGroupView>().from(view!!, mapBubble, onClickListener!!).root
                }
            }

            override fun updateView(mapBubble: MapBubble) {
                when (mapBubble.type) {
                    BubbleType.STATUS -> on<MapBubblePhysicalGroupView>().update(MapBubblePhysicalGroupBinding.bind(mapBubble.view!!), mapBubble)
                }
            }
        }

    fun attach(bubbleMapLayerLayout: ViewGroup,
               onClickListener: (MapBubble) -> Unit,
               onMenuItemClickListener: OnMapBubbleMenuItemClickListener,
               onMapBubbleEventClickListener: MapBubbleEventClickListener,
               onMapBubbleSuggestionClickListener: MapBubbleSuggestionClickListener,
               onMapBubblePhysicalGroupClickListener: MapBubblePhysicalGroupClickListener) {
        this.onClickListener = onClickListener
        this.onMenuItemClickListener = onMenuItemClickListener
        this.onMapBubbleEventClickListener = onMapBubbleEventClickListener
        this.onMapBubbleSuggestionClickListener = onMapBubbleSuggestionClickListener
        this.onMapBubblePhysicalGroupClickListener = onMapBubblePhysicalGroupClickListener
        bubbleMapLayer.attach(bubbleMapLayerLayout, bubbleView)
    }

    fun attach(map: Map) {
        bubbleMapLayer.attach(map)
    }

    fun add(mapBubble: MapBubble) {
        bubbleProxyLayer.add(mapBubble)
    }

    fun replace(mapBubbles: List<MapBubble>) {
        bubbleProxyLayer.replace(mapBubbles)
    }

    fun move(mapBubble: MapBubble, latLng: LatLng) {
        bubbleProxyLayer.move(mapBubble, latLng)
    }

    fun remove(mapBubble: MapBubble) {
        bubbleProxyLayer.remove(mapBubble)
    }

    fun remove(callback: (MapBubble) -> Boolean): Boolean {
        return bubbleProxyLayer.remove(callback)
    }

    fun update() {
        bubbleMapLayer.update()
    }

    fun update(mapBubble: MapBubble) {
        bubbleMapLayer.update(mapBubble)
    }

    fun updateDetails(mapBubble: MapBubble) {
        bubbleMapLayer.updateDetails(mapBubble)
    }
}
